### ISSUE TITLE
use_vt broken under freebsd

### ISSUE THREAD

**Issue by kim0** [saltstack/salt#20565]
## works
```
$ salt-call --local cmd.run cmd='uname -a'
[INFO    ] Executing command '/sbin/zfs -?' in directory '/root'
[INFO    ] Executing command '/sbin/zfs -?' in directory '/root'
[INFO    ] Executing command 'uname -a' in directory '/root'
local:
    FreeBSD salt.freebsd.org 10.1-RELEASE FreeBSD 10.1-RELEASE #0 r274401: Tue Nov 11 21:02:49 UTC 2014     root@releng1.nyi.freebsd.org:/usr/obj/usr/src/sys/GENERIC  amd64
```

## broken
```
$ salt-call --local cmd.run cmd='uname -a' use_vt=True
[INFO    ] Executing command '/sbin/zfs -?' in directory '/root'
[INFO    ] Executing command '/sbin/zfs -?' in directory '/root'
[INFO    ] Executing command 'uname -a' in directory '/root'
[WARNING ] Failed to spawn the VT: [Errno 6] Device not configured: '/dev/tty'
[ERROR   ] An un-handled exception was caught by salt's global exception handler:
UnboundLocalError: local variable 'proc' referenced before assignment
Traceback (most recent call last):
  File "/root/venv-salt/bin/salt-call", line 10, in <module>
    execfile(__file__)
  File "/root/salt/scripts/salt-call", line 11, in <module>
    salt_call()
  File "/root/salt/salt/scripts.py", line 220, in salt_call
    client.run()
  File "/root/salt/salt/cli/call.py", line 69, in run
    caller.run()
  File "/root/salt/salt/cli/caller.py", line 234, in run
    ret = self.call()
  File "/root/salt/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/root/salt/salt/modules/cmdmod.py", line 630, in run
    use_vt=use_vt)
  File "/root/salt/salt/modules/cmdmod.py", line 463, in _run
    proc.close(terminate=True, kill=True)
UnboundLocalError: local variable 'proc' referenced before assignment
Traceback (most recent call last):
  File "/root/venv-salt/bin/salt-call", line 10, in <module>
    execfile(__file__)
  File "/root/salt/scripts/salt-call", line 11, in <module>
    salt_call()
  File "/root/salt/salt/scripts.py", line 220, in salt_call
    client.run()
  File "/root/salt/salt/cli/call.py", line 69, in run
    caller.run()
  File "/root/salt/salt/cli/caller.py", line 234, in run
    ret = self.call()
  File "/root/salt/salt/cli/caller.py", line 138, in call
    ret['return'] = func(*args, **kwargs)
  File "/root/salt/salt/modules/cmdmod.py", line 630, in run
    use_vt=use_vt)
  File "/root/salt/salt/modules/cmdmod.py", line 463, in _run
    proc.close(terminate=True, kill=True)
UnboundLocalError: local variable 'proc' referenced before assignment
[ERROR   ] Command 'uname -a' failed with return code: 1
[ERROR   ] output:
```

**Comment by kim0** [comment]
Seems to me like the culprit is salt/utils/vt.py is broken on freebsd
This probably causes #18714 .. I spent sometime trying to debug it, but could't beat it .. All the pty kungfoo is too poorly documented

the utils.vt module has a re-implementation of __fork_pty and says this is done to provide separate stdout and stderr streams .. while I'm new to most of this pty stuff, I wonder if we could just use python's native implementation which returns a unix raw fileno which we might be able to os.dup2 to stdout and stderr streams.

Just wish someone who really understands this, takes a good look into it on freebsd. Thanks!

**Comment by rallytime** [comment]
@kim0 Thanks for the report! Can you also post the output of `salt --versions-report`?

**Comment by s0undt3ch** [comment]
Thanks for reporting this. VT supports interaction similar to a shell, something Python's subprocess does not support.

I'll try to get my hands on a BSD machine to debug this....

**Comment by kim0** [comment]
Salt: 2015.2.0-144-g3e2e9e3
         Python: 2.7.9 (default, Dec 31 2014, 03:40:40)
         Jinja2: 2.7.3
       M2Crypto: 0.22
 msgpack-python: 0.4.5
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
        libnacl: Not Installed
         PyYAML: 3.11
          ioflo: Not Installed
          PyZMQ: 14.5.0
           RAET: Not Installed
            ZMQ: 4.0.4
           Mako: Not Installed

**Comment by kim0** [comment]
@s0undt3ch I meant https://docs.python.org/2/library/pty.html

**Comment by s0undt3ch** [comment]
When developing the module I also investigated Pty, it didn't prove itself useful for what was needed.

**Comment by ndenev** [comment]
Out of desperation I've deleted locally all of the code between both comments *Make STDOUT the controlling PTY*, except for the os.setsid(). With this I was able to use cloud.saltify and cloud.digital_ocean to provision FreeBSD machines (however I've needed to patch cloud.py too, because  of the way shell commands are escaped was failing on FreeBSD's tcsh, but this is another story)

**Comment by cachedout** [comment]
Closed by #22120

**Comment by opsec** [comment]
Please have a look at

https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=204819
and
https://lists.freebsd.org/pipermail/freebsd-stable/2015-November/083710.html
and following posts.
